src/crypto/cn_slow_hash.hpp

```diff
@@ -0,0 +1,232 @@
+// THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Parts of this file are originally copyright (c) 2014-2017, The Monero Project
+// Parts of this file are originally copyright (c) 2012-2013, The Cryptonote developers
+
+#pragma once
+
+#include <inttypes.h>
+#include <stddef.h>
+#include <stdlib.h>
+#include <assert.h>
+#include <string.h>
+
+// Note HAS_INTEL_HW and future HAS_ARM_HW only mean we can emit the AES instructions
+// check CPU support for the hardware AES encryption has to be done at runtime
+#if defined(__x86_64__) || defined(__i386__) || defined(_M_X86) || defined(_M_X64)
+#ifdef __GNUC__
+#include <x86intrin.h>
+#include <cpuid.h>
+#pragma GCC target ("aes")
+#define HAS_INTEL_HW
+#endif
+#ifdef _MSC_VER 
+#include <intrin.h>
+#define HAS_INTEL_HW
+#endif
+#endif
+
+#if defined(_WIN32) || defined(_WIN64)
+#include <malloc.h>
+#define WIN_MEM_ALIGN
+#endif
+
+#ifdef HAS_INTEL_HW
+inline void cpuid(uint32_t eax, int32_t ecx, int32_t val[4])
+{
+	val[0] = 0;
+	val[1] = 0;
+	val[2] = 0;
+	val[3] = 0;
+
+#ifdef _MSC_VER
+	__cpuidex(val, eax, ecx);
+#else
+	__cpuid_count(eax, ecx, val[0], val[1], val[2], val[3]);
+#endif
+}
+
+inline bool hw_check_aes()
+{
+	int32_t cpu_info[4];
+	cpuid(1, 0, cpu_info);
+	return (cpu_info[2] & (1 << 25)) != 0;
+}
+#endif
+
+#ifdef HAS_ARM_HW
+inline bool hw_check_aes()
+{
+	return false;
+}
+#endif
+
+#if !defined(HAS_INTEL_HW) && !defined(HAS_ARM_HW)
+inline bool hw_check_aes()
+{
+	return false;
+}
+#endif
+
+// This cruft avoids casting-galore and allows us not to worry about sizeof(void*)
+union cn_sptr
+{
+	cn_sptr() : as_void(nullptr) {}
+	cn_sptr(uint64_t* ptr) { as_uqword = ptr; }
+	cn_sptr(uint32_t* ptr) { as_udword = ptr; }
+	cn_sptr(uint8_t* ptr) { as_byte = ptr; }
+#ifdef HAS_INTEL_HW
+	cn_sptr(__m128i* ptr) { as_xmm = ptr; }
+#endif
+
+	void* as_void;
+	uint8_t* as_byte;
+	uint64_t* as_uqword;
+	int32_t* as_dword;
+	uint32_t* as_udword;
+#ifdef HAS_INTEL_HW
+	__m128i* as_xmm;
+#endif
+};
+
+#include <stdio.h>
+#include <signal.h>
+template<size_t MEMORY, size_t ITER>
+class cn_slow_hash
+{
+public:
+	cn_slow_hash()
+	{
+#if !defined(WIN_MEM_ALIGN)
+		lpad.as_void = aligned_alloc(4096, MEMORY);
+		spad.as_void = aligned_alloc(4096, 4096);
+#else
+		lpad.as_void = _aligned_malloc(MEMORY, 4096);
+		spad.as_void = _aligned_malloc(4096, 4096);
+#endif
+		printf("cn_slow_hash allocated %.16lx %.16lx %.16lx!\n", uint64_t(this), uint64_t(lpad.as_void), uint64_t(spad.as_void));
```

Comment:
Less warning on MingW and other x86 platforms while compiling with:

`printf("cn_slow_hash allocated %.16llx %.16llx %.16llx!\n", uint64_t(this), uint64_t(lpad.as_void), uint64_t(spad.as_void));`

Is it for debug only?

Reply:
Yes. Sorry for pushing it.